src/core/Akka.Streams/Stage/GraphStage.cs

```diff
@@ -1683,8 +1695,77 @@ protected Action<T> GetAsyncCallback<T>(Action<T> handler)
         /// <param name="handler">TBD</param>
         /// <returns>TBD</returns>
         protected Action GetAsyncCallback(Action handler)
-            => () => Interpreter.OnAsyncInput(this, NotUsed.Instance, _ => handler());
+            => () => Interpreter.OnAsyncInput(this, NotUsed.Instance, NoPromise, _ => handler());
 
+        protected Func<T, Task<Done>> GetAsyncCallbackAsync<T>(Action<T> handler)
+        {
+            return @event =>
+            {
+                var promise = new TaskCompletionSource<Done>();
+                if (AddToWaiting(promise))
+                {
+                    Interpreter.OnAsyncInput(this, @event, promise, x => handler((T)x));
+                }
```

Comment:
We create a new `TaskCompletionSource` "promise" for each async callback invocation. `AddToWaiting()` checks that the promise can be atomically added to `AsyncCallbacksInProgress`. If the add succeed, we enqueue its invocation in the interpreter, if not, we signal a `StreamDetachedException` on the promise.